File: web/src/services/localDatabase.ts
import initSqlJs from 'sql.js';
import { get, set, del } from 'idb-keyval';
import sqliteWasmUrl from 'sql.js/dist/sql-wasm.wasm?url';
import type { Project, Task } from '../types/api';

const DB_KEY = 'atesteme-taskmanager-sqlite';

type Database = initSqlJs.Database;
type SqlJsStatic = initSqlJs.SqlJsStatic;
type SqlValue = initSqlJs.SqlValue;
type QueryRow = Record<string, SqlValue>;

export type PendingOperationType =
  | 'CREATE_PROJECT'
  | 'UPDATE_PROJECT'
  | 'DELETE_PROJECT'
  | 'CREATE_TASK'
  | 'UPDATE_TASK'
  | 'UPDATE_TASK_STATUS'
  | 'DELETE_TASK';

export type PendingOperation = {
  id: number;
  operationType: PendingOperationType;
  resourceId: number | null;
  projectId: number | null;
  payload: Record<string, unknown> | null;
  createdAt: string;
};

class LocalDatabase {
  private databasePromise: Promise<Database> | null = null;
  private sqlPromise: Promise<SqlJsStatic> | null = null;

  async saveProjects(projects: Project[]) {
    const db = await this.database();
    db.run('DELETE FROM projects');

    const statement = db.prepare(`
      INSERT INTO projects (id, name, description, workers, created_at, task_count)
      VALUES (?, ?, ?, ?, ?, ?)
    `);

    try {
      projects.forEach((project) => {
        statement.run([
          project.id,
          project.name,
          project.description ?? null,
          JSON.stringify(project.workers ?? []),
          project.createdAt,
          project.taskCount
        ]);
      });
    } finally {
      statement.free();
    }

    await this.persist();
  }

  async upsertProject(project: Project) {
    const db = await this.database();
    db.run(`
      INSERT INTO projects (id, name, description, workers, created_at, task_count)
      VALUES (?, ?, ?, ?, ?, ?)
      ON CONFLICT(id) DO UPDATE SET
        name = excluded.name,
        description = excluded.description,
        workers = excluded.workers,
        created_at = excluded.created_at,
        task_count = excluded.task_count
    `, [
      project.id,
      project.name,
      project.description ?? null,
      JSON.stringify(project.workers ?? []),
      project.createdAt,
      project.taskCount
    ]);
    await this.persist();
  }

  async deleteProject(projectId: number) {
    const db = await this.database();
    db.run('DELETE FROM tasks WHERE project_id = ?', [projectId]);
    db.run('DELETE FROM projects WHERE id = ?', [projectId]);
    await this.persist();
  }

  async getProjects() {
    const db = await this.database();
    return this.query<Project>(db, `
      SELECT id, name, description, workers, created_at, task_count
      FROM projects
      ORDER BY datetime(created_at) DESC
    `, (row) => ({
      id: Number(row.id),
      name: String(row.name),
      description: row.description === null ? null : String(row.description),
      workers: this.parseWorkers(row.workers),
      createdAt: String(row.created_at),
      taskCount: Number(row.task_count)
    }));
  }

  async getProject(projectId: number) {
    const projects = await this.getProjects();
    return projects.find((project) => project.id === projectId) ?? null;
  }

  async saveTasks(projectId: number, tasks: Task[]) {
    const db = await this.database();
    db.run('DELETE FROM tasks WHERE project_id = ?', [projectId]);

    const statement = db.prepare(`
      INSERT INTO tasks (id, project_id, title, description, priority, status, created_at)
      VALUES (?, ?, ?, ?, ?, ?, ?)
    `);

    try {
      tasks.forEach((task) => {
        statement.run([
          task.id,
          task.projectId,
          task.title,
          task.description ?? null,
          task.priority,
          task.status,
          task.createdAt
        ]);
      });
    } finally {
      statement.free();
    }

    await this.persist();
  }

  async upsertTask(task: Task) {
    const db = await this.database();
    db.run(`
      INSERT INTO tasks (id, project_id, title, description, priority, status, created_at)
      VALUES (?, ?, ?, ?, ?, ?, ?)
      ON CONFLICT(id) DO UPDATE SET
        project_id = excluded.project_id,
        title = excluded.title,
        description = excluded.description,
        priority = excluded.priority,
        status = excluded.status,
        created_at = excluded.created_at
    `, [
      task.id,
      task.projectId,
      task.title,
      task.description ?? null,
      task.priority,
      task.status,
      task.createdAt
    ]);
    await this.persist();
  }

  async deleteTask(taskId: number) {
    const db = await this.database();
    db.run('DELETE FROM tasks WHERE id = ?', [taskId]);
    await this.persist();
  }

  async getTasks(projectId: number) {
    const db = await this.database();
    return this.query<Task>(db, `
      SELECT id, project_id, title, description, priority, status, created_at
      FROM tasks
      WHERE project_id = ?
      ORDER BY datetime(created_at) DESC
    `, (row) => ({
      id: Number(row.id),
      projectId: Number(row.project_id),
      title: String(row.title),
      description: row.description === null ? null : String(row.description),
      priority: row.priority as Task['priority'],
      status: row.status as Task['status'],
      createdAt: String(row.created_at)
    }), [projectId]);
  }

  async getTask(projectId: number, taskId: number) {
    const tasks = await this.getTasks(projectId);
    return tasks.find((task) => task.id === taskId) ?? null;
  }

  async addPendingOperation(
    operationType: PendingOperationType,
    payload: Record<string, unknown> | null,
    options: { resourceId?: number; projectId?: number } = {}
  ) {
    const db = await this.database();
    this.resolvePendingOperation(db, operationType, payload, options);
    await this.persist();
  }

  async getPendingOperations() {
    const db = await this.database();
    return this.query<PendingOperation>(db, `
      SELECT id, operation_type, resource_id, project_id, payload, created_at
      FROM pending_operations
      ORDER BY id ASC
    `, (row) => ({
      id: Number(row.id),
      operationType: row.operation_type as PendingOperationType,
      resourceId: row.resource_id === null ? null : Number(row.resource_id),
      projectId: row.project_id === null ? null : Number(row.project_id),
      payload: row.payload === null ? null : JSON.parse(String(row.payload)) as Record<string, unknown>,
      createdAt: String(row.created_at)
    }));
  }

  async deletePendingOperation(operationId: number) {
    const db = await this.database();
    db.run('DELETE FROM pending_operations WHERE id = ?', [operationId]);
    await this.persist();
  }

  async deletePendingOperationsForResource(resourceId: number) {
    const db = await this.database();
    db.run('DELETE FROM pending_operations WHERE resource_id = ?', [resourceId]);
    await this.persist();
  }

  async deletePendingOperationsForProject(projectId: number) {
    const db = await this.database();
    db.run('DELETE FROM pending_operations WHERE resource_id = ? OR project_id = ?', [projectId, projectId]);
    await this.persist();
  }

  async replaceProjectReferences(fromProjectId: number, toProjectId: number) {
    const db = await this.database();
    db.run('UPDATE tasks SET project_id = ? WHERE project_id = ?', [toProjectId, fromProjectId]);
    db.run('UPDATE pending_operations SET project_id = ? WHERE project_id = ?', [toProjectId, fromProjectId]);
    db.run(`
      UPDATE pending_operations
      SET resource_id = ?
      WHERE resource_id = ?
        AND operation_type IN ('UPDATE_PROJECT', 'DELETE_PROJECT')
    `, [toProjectId, fromProjectId]);
    await this.persist();
  }

  async clearAll() {
    const db = await this.database();
    db.run('DELETE FROM tasks');
    db.run('DELETE FROM projects');
    db.run('DELETE FROM pending_operations');
    await this.persist();
    await del(DB_KEY);
  }

  private async database() {
    if (!this.databasePromise) {
      this.databasePromise = this.createDatabase();
    }

    return this.databasePromise;
  }

  private async createDatabase() {
    const SQL = await this.sql();
    const persisted = await get<Uint8Array>(DB_KEY);
    const db = persisted ? new SQL.Database(persisted) : new SQL.Database();
    this.createSchema(db);
    return db;
  }

  private async sql() {
    if (!this.sqlPromise) {
      this.sqlPromise = initSqlJs({
        locateFile: () => sqliteWasmUrl
      });
    }

    return this.sqlPromise;
  }

  private createSchema(db: Database) {
    db.run(`
      CREATE TABLE IF NOT EXISTS projects (
        id INTEGER PRIMARY KEY,
        name TEXT NOT NULL,
        description TEXT,
        workers TEXT,
        created_at TEXT NOT NULL,
        task_count INTEGER NOT NULL DEFAULT 0
      );

      CREATE TABLE IF NOT EXISTS tasks (
        id INTEGER PRIMARY KEY,
        project_id INTEGER NOT NULL,
        title TEXT NOT NULL,
        description TEXT,
        priority TEXT NOT NULL,
        status TEXT NOT NULL,
        created_at TEXT NOT NULL
      );

      CREATE INDEX IF NOT EXISTS idx_tasks_project_id ON tasks(project_id);

      CREATE TABLE IF NOT EXISTS pending_operations (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        operation_type TEXT NOT NULL,
        resource_id INTEGER,
        project_id INTEGER,
        payload TEXT,
        created_at TEXT NOT NULL
      );
    `);
    this.ensureColumn(db, 'projects', 'workers', 'TEXT');
  }

  private ensureColumn(db: Database, table: string, column: string, definition: string) {
    const columns = this.query<string>(
      db,
      `PRAGMA table_info(${table})`,
      (row) => String(row.name)
    );

    if (!columns.includes(column)) {
      db.run(`ALTER TABLE ${table} ADD COLUMN ${column} ${definition}`);
    }
  }

  private parseWorkers(value: SqlValue) {
    if (value === null) {
      return [];
    }

    try {
      const parsed = JSON.parse(String(value)) as unknown;
      return Array.isArray(parsed)
        ? parsed.filter((worker): worker is string => typeof worker === 'string')
        : [];
    } catch {
      return [];
    }
  }

  private resolvePendingOperation(
    db: Database,
    operationType: PendingOperationType,
    payload: Record<string, unknown> | null,
    options: { resourceId?: number; projectId?: number }
  ) {
    const resourceId = options.resourceId ?? null;
    const projectId = options.projectId ?? null;

    if (resourceId === null) {
      this.insertPendingOperation(db, operationType, payload, resourceId, projectId);
      return;
    }

    const existing = this.pendingOperationsForResource(db, operationType, resourceId, projectId);
    const existingDelete = existing.find((operation) => this.isDeleteOperation(operation.operationType));

    if (existingDelete && !this.isDeleteOperation(operationType)) {
      return;
    }

    if (this.isDeleteOperation(operationType)) {
      const projectTaskOperations = operationType === 'DELETE_PROJECT'
        ? this.pendingTaskOperationsForProject(db, resourceId)
        : [];
      const operationsToRemove = [...existing, ...projectTaskOperations];
      const pendingCreate = existing.find((operation) => this.isCreateOperation(operation.operationType));

      this.deletePendingOperationsByIds(db, operationsToRemove.map((operation) => operation.id));

      if (pendingCreate || resourceId < 0) {
        return;
      }

      this.insertPendingOperation(db, operationType, payload, resourceId, projectId);
      return;
    }

    const pendingCreate = existing.find((operation) => this.isCreateOperation(operation.operationType));
    if (pendingCreate) {
      this.updatePendingOperationPayload(
        db,
        pendingCreate.id,
        this.mergePayload(pendingCreate.payload, payload, operationType)
      );
      this.deletePendingOperationsByIds(
        db,
        existing
          .filter((operation) => operation.id !== pendingCreate.id)
          .map((operation) => operation.id)
      );
      return;
    }

    if (operationType === 'UPDATE_TASK_STATUS') {
      const pendingTaskUpdate = existing.find((operation) => operation.operationType === 'UPDATE_TASK');
      if (pendingTaskUpdate) {
        this.updatePendingOperationPayload(
          db,
          pendingTaskUpdate.id,
          this.mergePayload(pendingTaskUpdate.payload, payload, operationType)
        );
        this.deletePendingOperationsByIds(
          db,
          existing
            .filter((operation) => operation.operationType === 'UPDATE_TASK_STATUS')
            .map((operation) => operation.id)
        );
        return;
      }
    }

    const replaceableTypes = this.replaceableOperationTypes(operationType);
    this.deletePendingOperationsByIds(
      db,
      existing
        .filter((operation) => replaceableTypes.includes(operation.operationType))
        .map((operation) => operation.id)
    );
    this.insertPendingOperation(db, operationType, payload, resourceId, projectId);
  }

  private insertPendingOperation(
    db: Database,
    operationType: PendingOperationType,
    payload: Record<string, unknown> | null,
    resourceId: number | null,
    projectId: number | null
  ) {
    db.run(`
      INSERT INTO pending_operations (operation_type, resource_id, project_id, payload, created_at)
      VALUES (?, ?, ?, ?, ?)
    `, [
      operationType,
      resourceId,
      projectId,
      payload ? JSON.stringify(payload) : null,
      new Date().toISOString()
    ]);
  }

  private updatePendingOperationPayload(
    db: Database,
    operationId: number,
    payload: Record<string, unknown> | null
  ) {
    db.run('UPDATE pending_operations SET payload = ?, created_at = ? WHERE id = ?', [
      payload ? JSON.stringify(payload) : null,
      new Date().toISOString(),
      operationId
    ]);
  }

  private deletePendingOperationsByIds(db: Database, operationIds: number[]) {
    [...new Set(operationIds)].forEach((operationId) => {
      db.run('DELETE FROM pending_operations WHERE id = ?', [operationId]);
    });
  }

  private pendingOperationsForResource(
    db: Database,
    operationType: PendingOperationType,
    resourceId: number,
    projectId: number | null
  ) {
    return this.query<PendingOperation>(db, `
      SELECT id, operation_type, resource_id, project_id, payload, created_at
      FROM pending_operations
      WHERE resource_id = ?
      ORDER BY id ASC
    `, (row) => this.mapPendingOperation(row), [resourceId])
      .filter((operation) => {
        if (this.isProjectOperation(operationType) !== this.isProjectOperation(operation.operationType)) {
          return false;
        }

        return this.isProjectOperation(operationType)
          || projectId === null
          || operation.projectId === projectId;
      });
  }

  private pendingTaskOperationsForProject(db: Database, projectId: number) {
    return this.query<PendingOperation>(db, `
      SELECT id, operation_type, resource_id, project_id, payload, created_at
      FROM pending_operations
      WHERE project_id = ?
      ORDER BY id ASC
    `, (row) => this.mapPendingOperation(row), [projectId])
      .filter((operation) => this.isTaskOperation(operation.operationType));
  }

  private replaceableOperationTypes(operationType: PendingOperationType) {
    switch (operationType) {
      case 'UPDATE_PROJECT':
        return ['UPDATE_PROJECT'];
      case 'UPDATE_TASK':
        return ['UPDATE_TASK', 'UPDATE_TASK_STATUS'];
      case 'UPDATE_TASK_STATUS':
        return ['UPDATE_TASK_STATUS'];
      default:
        return [operationType];
    }
  }

  private mergePayload(
    current: Record<string, unknown> | null,
    next: Record<string, unknown> | null,
    operationType: PendingOperationType
  ) {
    const merged = { ...(current ?? {}) };

    if (operationType === 'UPDATE_TASK_STATUS') {
      if (next && 'status' in next) {
        merged.status = next.status;
      }

      return Object.keys(merged).length ? merged : next;
    }

    return {
      ...merged,
      ...(next ?? {})
    };
  }

  private mapPendingOperation(row: QueryRow): PendingOperation {
    return {
      id: Number(row.id),
      operationType: row.operation_type as PendingOperationType,
      resourceId: row.resource_id === null ? null : Number(row.resource_id),
      projectId: row.project_id === null ? null : Number(row.project_id),
      payload: row.payload === null ? null : JSON.parse(String(row.payload)) as Record<string, unknown>,
      createdAt: String(row.created_at)
    };
  }

  private isCreateOperation(operationType: PendingOperationType) {
    return operationType === 'CREATE_PROJECT' || operationType === 'CREATE_TASK';
  }

  private isDeleteOperation(operationType: PendingOperationType) {
    return operationType === 'DELETE_PROJECT' || operationType === 'DELETE_TASK';
  }

  private isProjectOperation(operationType: PendingOperationType) {
    return operationType.endsWith('_PROJECT');
  }

  private isTaskOperation(operationType: PendingOperationType) {
    return !this.isProjectOperation(operationType);
  }

  private async persist() {
    const db = await this.database();
    // O banco local é cache de UX/offline; o backend segue como fonte das regras sensíveis.
    await set(DB_KEY, db.export());
  }

  private query<T>(
    db: Database,
    sql: string,
    mapper: (row: QueryRow) => T,
    params: SqlValue[] = []
  ) {
    const statement = db.prepare(sql, params);
    const rows: T[] = [];

    try {
      while (statement.step()) {
        rows.push(mapper(statement.getAsObject()));
      }
    } finally {
      statement.free();
    }

    return rows;
  }
}

export const localDatabase = new LocalDatabase();

File: mobile/lib/app/shared/storage/local_database.dart
import 'dart:convert';

import 'package:path/path.dart' as p;
import 'package:sqflite/sqflite.dart';

import '../../modules/projects/project_model.dart';
import '../../modules/tasks/task_model.dart';

class LocalDatabase {
  Database? _database;

  Future<void> upsertProject(ProjectModel project) async {
    final db = await database;
    await db.insert('projects', _projectToMap(project), conflictAlgorithm: ConflictAlgorithm.replace);
  }

  Future<void> deleteProject(int projectId) async {
    final db = await database;
    await db.transaction((txn) async {
      await txn.delete('tasks', where: 'project_id = ?', whereArgs: [projectId]);
      await txn.delete('projects', where: 'id = ?', whereArgs: [projectId]);
    });
  }

  Future<void> saveProjects(List<ProjectModel> projects) async {
    final db = await database;
    await db.transaction((txn) async {
      await txn.delete('projects');
      for (final project in projects) {
        await txn.insert('projects', _projectToMap(project), conflictAlgorithm: ConflictAlgorithm.replace);
      }
    });
  }

  Future<List<ProjectModel>> getProjects() async {
    final db = await database;
    final rows = await db.query('projects', orderBy: 'created_at DESC');
    return rows.map(ProjectModel.fromDatabase).toList();
  }

  Future<ProjectModel?> getProject(int projectId) async {
    final db = await database;
    final rows = await db.query('projects', where: 'id = ?', whereArgs: [projectId], limit: 1);
    if (rows.isEmpty) {
      return null;
    }

    return ProjectModel.fromDatabase(rows.first);
  }

  Future<void> saveTasks(int projectId, List<TaskModel> tasks) async {
    final db = await database;
    await db.transaction((txn) async {
      await txn.delete('tasks', where: 'project_id = ?', whereArgs: [projectId]);
      for (final task in tasks) {
        await txn.insert('tasks', _taskToMap(task), conflictAlgorithm: ConflictAlgorithm.replace);
      }
    });
  }

  Future<List<TaskModel>> getTasks(int projectId) async {
    final db = await database;
    final rows = await db.query(
      'tasks',
      where: 'project_id = ?',
      whereArgs: [projectId],
      orderBy: 'created_at DESC',
    );
    return rows.map(TaskModel.fromDatabase).toList();
  }

  Future<void> upsertTask(TaskModel task) async {
    final db = await database;
    await db.insert('tasks', _taskToMap(task), conflictAlgorithm: ConflictAlgorithm.replace);
  }

  Future<void> deleteTask(int taskId) async {
    final db = await database;
    await db.delete('tasks', where: 'id = ?', whereArgs: [taskId]);
  }

  Future<void> addPendingOperation({
    required String operationType,
    required Map<String, dynamic>? payload,
    int? resourceId,
    int? projectId,
  }) async {
    final db = await database;
    await _resolvePendingOperation(
      db,
      operationType: operationType,
      payload: payload,
      resourceId: resourceId,
      projectId: projectId,
    );
  }

  Future<List<Map<String, Object?>>> getPendingOperations() async {
    final db = await database;
    return db.query('pending_operations', orderBy: 'id ASC');
  }

  Future<void> deletePendingOperation(int operationId) async {
    final db = await database;
    await db.delete('pending_operations', where: 'id = ?', whereArgs: [operationId]);
  }

  Future<void> deletePendingOperationsForResource(int resourceId) async {
    final db = await database;
    await db.delete('pending_operations', where: 'resource_id = ?', whereArgs: [resourceId]);
  }

  Future<void> deletePendingOperationsForProject(int projectId) async {
    final db = await database;
    await db.delete(
      'pending_operations',
      where: 'resource_id = ? OR project_id = ?',
      whereArgs: [projectId, projectId],
    );
  }

  Future<void> replaceProjectReferences({
    required int fromProjectId,
    required int toProjectId,
  }) async {
    final db = await database;
    await db.transaction((txn) async {
      await txn.update(
        'tasks',
        {'project_id': toProjectId},
        where: 'project_id = ?',
        whereArgs: [fromProjectId],
      );
      await txn.update(
        'pending_operations',
        {'project_id': toProjectId},
        where: 'project_id = ?',
        whereArgs: [fromProjectId],
      );
      await txn.update(
        'pending_operations',
        {'resource_id': toProjectId},
        where: "resource_id = ? AND operation_type IN ('UPDATE_PROJECT', 'DELETE_PROJECT')",
        whereArgs: [fromProjectId],
      );
    });
  }

  Future<void> clearPendingOperations() async {
    final db = await database;
    await db.delete('pending_operations');
  }

  Future<void> clearAll() async {
    final db = await database;
    await db.delete('tasks');
    await db.delete('projects');
    await db.delete('pending_operations');
  }

  Future<Database> get database async {
    if (_database != null) {
      return _database!;
    }

    final databasePath = await getDatabasesPath();
    final path = p.join(databasePath, 'atesteme_taskmanager.db');
    _database = await openDatabase(
      path,
      version: 3,
      onCreate: (db, _) async {
        await db.execute('''
          CREATE TABLE projects (
            id INTEGER PRIMARY KEY,
            name TEXT NOT NULL,
            description TEXT,
            workers TEXT,
            created_at TEXT NOT NULL,
            task_count INTEGER NOT NULL DEFAULT 0
          )
        ''');

        await db.execute('''
          CREATE TABLE tasks (
            id INTEGER PRIMARY KEY,
            project_id INTEGER NOT NULL,
            title TEXT NOT NULL,
            description TEXT,
            priority TEXT NOT NULL,
            status TEXT NOT NULL,
            created_at TEXT NOT NULL
          )
        ''');

        await db.execute('CREATE INDEX idx_tasks_project_id ON tasks(project_id)');

        await db.execute('''
          CREATE TABLE pending_operations (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            operation_type TEXT NOT NULL,
            resource_id INTEGER,
            project_id INTEGER,
            payload TEXT,
            created_at TEXT NOT NULL
          )
        ''');
      },
      onUpgrade: (db, oldVersion, _) async {
        if (oldVersion < 2) {
          await db.execute('''
            CREATE TABLE IF NOT EXISTS pending_operations (
              id INTEGER PRIMARY KEY AUTOINCREMENT,
              operation_type TEXT NOT NULL,
              resource_id INTEGER,
              project_id INTEGER,
              payload TEXT,
              created_at TEXT NOT NULL
            )
          ''');
        }
        if (oldVersion < 3) {
          await db.execute('ALTER TABLE projects ADD COLUMN workers TEXT');
        }
      },
    );

    return _database!;
  }

  Map<String, Object?> _projectToMap(ProjectModel project) {
    return {
      'id': project.id,
      'name': project.name,
      'description': project.description,
      'workers': project.workers.join('\n'),
      'created_at': project.createdAt.toIso8601String(),
      'task_count': project.taskCount,
    };
  }

  Map<String, Object?> _taskToMap(TaskModel task) {
    return {
      'id': task.id,
      'project_id': task.projectId,
      'title': task.title,
      'description': task.description,
      'priority': task.priority.value,
      'status': task.status.value,
      'created_at': task.createdAt.toIso8601String(),
    };
  }

  Future<void> _resolvePendingOperation(
    Database db, {
    required String operationType,
    required Map<String, dynamic>? payload,
    required int? resourceId,
    required int? projectId,
  }) async {
    if (resourceId == null) {
      await _insertPendingOperation(
        db,
        operationType: operationType,
        payload: payload,
        resourceId: resourceId,
        projectId: projectId,
      );
      return;
    }

    final existing = await _pendingOperationsForResource(
      db,
      operationType: operationType,
      resourceId: resourceId,
      projectId: projectId,
    );
    final existingDelete = _firstWhereOrNull(existing, (operation) {
      return _isDeleteOperation(operation['operation_type'] as String);
    });

    if (existingDelete != null && !_isDeleteOperation(operationType)) {
      return;
    }

    if (_isDeleteOperation(operationType)) {
      final projectTaskOperations = operationType == 'DELETE_PROJECT'
          ? await _pendingTaskOperationsForProject(db, resourceId)
          : <Map<String, Object?>>[];
      final pendingCreate = _firstWhereOrNull(existing, (operation) {
        return _isCreateOperation(operation['operation_type'] as String);
      });

      await _deletePendingOperationIds(
        db,
        [
          ...existing,
          ...projectTaskOperations,
        ].map((operation) => operation['id'] as int),
      );

      if (pendingCreate != null || resourceId < 0) {
        return;
      }

      await _insertPendingOperation(
        db,
        operationType: operationType,
        payload: payload,
        resourceId: resourceId,
        projectId: projectId,
      );
      return;
    }

    final pendingCreate = _firstWhereOrNull(existing, (operation) {
      return _isCreateOperation(operation['operation_type'] as String);
    });
    if (pendingCreate != null) {
      await _updatePendingOperationPayload(
        db,
        pendingCreate['id'] as int,
        _mergePayload(
          decodeOperationPayload(pendingCreate['payload']),
          payload,
          operationType,
        ),
      );
      await _deletePendingOperationIds(
        db,
        existing
            .where((operation) => operation['id'] != pendingCreate['id'])
            .map((operation) => operation['id'] as int),
      );
      return;
    }

    if (operationType == 'UPDATE_TASK_STATUS') {
      final pendingTaskUpdate = _firstWhereOrNull(existing, (operation) {
        return operation['operation_type'] == 'UPDATE_TASK';
      });
      if (pendingTaskUpdate != null) {
        await _updatePendingOperationPayload(
          db,
          pendingTaskUpdate['id'] as int,
          _mergePayload(
            decodeOperationPayload(pendingTaskUpdate['payload']),
            payload,
            operationType,
          ),
        );
        await _deletePendingOperationIds(
          db,
          existing
              .where((operation) => operation['operation_type'] == 'UPDATE_TASK_STATUS')
              .map((operation) => operation['id'] as int),
        );
        return;
      }
    }

    final replaceableTypes = _replaceableOperationTypes(operationType);
    await _deletePendingOperationIds(
      db,
      existing
          .where((operation) => replaceableTypes.contains(operation['operation_type']))
          .map((operation) => operation['id'] as int),
    );
    await _insertPendingOperation(
      db,
      operationType: operationType,
      payload: payload,
      resourceId: resourceId,
      projectId: projectId,
    );
  }

  Future<void> _insertPendingOperation(
    Database db, {
    required String operationType,
    required Map<String, dynamic>? payload,
    required int? resourceId,
    required int? projectId,
  }) async {
    await db.insert('pending_operations', {
      'operation_type': operationType,
      'resource_id': resourceId,
      'project_id': projectId,
      'payload': payload == null ? null : jsonEncode(payload),
      'created_at': DateTime.now().toIso8601String(),
    });
  }

  Future<void> _updatePendingOperationPayload(
    Database db,
    int operationId,
    Map<String, dynamic>? payload,
  ) async {
    await db.update(
      'pending_operations',
      {
        'payload': payload == null ? null : jsonEncode(payload),
        'created_at': DateTime.now().toIso8601String(),
      },
      where: 'id = ?',
      whereArgs: [operationId],
    );
  }

  Future<void> _deletePendingOperationIds(Database db, Iterable<int> operationIds) async {
    for (final operationId in operationIds.toSet()) {
      await db.delete('pending_operations', where: 'id = ?', whereArgs: [operationId]);
    }
  }

  Future<List<Map<String, Object?>>> _pendingOperationsForResource(
    Database db, {
    required String operationType,
    required int resourceId,
    required int? projectId,
  }) async {
    final rows = await db.query(
      'pending_operations',
      where: 'resource_id = ?',
      whereArgs: [resourceId],
      orderBy: 'id ASC',
    );

    return rows.where((operation) {
      final existingType = operation['operation_type'] as String;
      if (_isProjectOperation(operationType) != _isProjectOperation(existingType)) {
        return false;
      }

      return _isProjectOperation(operationType) || projectId == null || operation['project_id'] == projectId;
    }).toList();
  }

  Future<List<Map<String, Object?>>> _pendingTaskOperationsForProject(Database db, int projectId) async {
    final rows = await db.query(
      'pending_operations',
      where: 'project_id = ?',
      whereArgs: [projectId],
      orderBy: 'id ASC',
    );

    return rows.where((operation) {
      return _isTaskOperation(operation['operation_type'] as String);
    }).toList();
  }

  List<String> _replaceableOperationTypes(String operationType) {
    switch (operationType) {
      case 'UPDATE_PROJECT':
        return ['UPDATE_PROJECT'];
      case 'UPDATE_TASK':
        return ['UPDATE_TASK', 'UPDATE_TASK_STATUS'];
      case 'UPDATE_TASK_STATUS':
        return ['UPDATE_TASK_STATUS'];
      default:
        return [operationType];
    }
  }

  Map<String, dynamic>? _mergePayload(
    Map<String, dynamic>? current,
    Map<String, dynamic>? next,
    String operationType,
  ) {
    final merged = <String, dynamic>{
      if (current != null) ...current,
    };

    if (operationType == 'UPDATE_TASK_STATUS') {
      if (next != null && next.containsKey('status')) {
        merged['status'] = next['status'];
      }

      return merged.isEmpty ? next : merged;
    }

    if (next != null) {
      merged.addAll(next);
    }

    return merged.isEmpty ? null : merged;
  }

  Map<String, Object?>? _firstWhereOrNull(
    Iterable<Map<String, Object?>> operations,
    bool Function(Map<String, Object?> operation) test,
  ) {
    for (final operation in operations) {
      if (test(operation)) {
        return operation;
      }
    }

    return null;
  }

  bool _isCreateOperation(String operationType) {
    return operationType == 'CREATE_PROJECT' || operationType == 'CREATE_TASK';
  }

  bool _isDeleteOperation(String operationType) {
    return operationType == 'DELETE_PROJECT' || operationType == 'DELETE_TASK';
  }

  bool _isProjectOperation(String operationType) {
    return operationType.endsWith('_PROJECT');
  }

  bool _isTaskOperation(String operationType) {
    return !_isProjectOperation(operationType);
  }
}

Map<String, dynamic>? decodeOperationPayload(Object? payload) {
  if (payload == null) {
    return null;
  }

  return jsonDecode(payload as String) as Map<String, dynamic>;
}
